src/os/bluestore/BlueStore.cc

```diff
@@ -5363,706 +5225,466 @@ void BlueStore::_pad_zeros_tail(
-	cow_head_op->src_extent.offset =
-	  b->calc_offset(x_off + bp->second.offset, &c_len);
-	c_len = MIN(c_len, cow_length);
-	cow_head_op->src_extent.length = c_len;
-	cow_head_op->extent.offset = 0;   // will reset this below
-	cow_head_op->extent.length = c_len;
-	x_off += c_len;
-	cow_length -= c_len;
-	assert(cow_length == 0);  // fixme ... until we fix this loop
+    // read some data to fill out the chunk?
+    uint64_t head_read = b_off % chunk_size;
+    uint64_t tail_read =
+      ROUND_UP_TO(b_off + b_len, chunk_size) - (b_off + b_len);
+    if ((head_read || tail_read) &&
+	b->get_ondisk_length() >= b_off + b_len + head_read + tail_read) {
+      dout(20) << __func__ << "  reading head 0x" << std::hex << head_read
+	       << " and tail 0x" << tail_read << std::dec << dendl;
+      if (head_read) {
+	bufferlist head_bl;
+	int r = _do_read(c.get(), o, offset - head_pad - head_read, head_read,
+			 head_bl, 0);
+	assert(r >= 0 && r <= (int)head_read);
+	size_t zlen = head_read - r;
+	if (zlen) {
+	  head_bl.append_zero(zlen);
+	}
+	b_off -= head_read;
+	b_len += head_read;
+	head_bl.claim_append(padded);
+	padded.swap(head_bl);
       }
-    } else {
-      dout(20) << "  head shared, but no COW needed" << dendl;
-    }
-    if (orig_offset & ~block_mask) {
-      *cow_rmw_head = b->calc_offset(orig_offset - bp->first, nullptr);
-      dout(20) << "  cow_rmw_head 0x" << std::hex << *cow_rmw_head << std::dec
-	       << " from " << bp->second
-	       << " blob " << *b << dendl;
-    }
-  }
-  if (shared_tail) {
-    uint64_t cow_offset_raw = orig_offset + orig_length;
-    uint64_t cow_offset = ROUND_UP_TO(cow_offset_raw, block_size);
-    uint64_t cow_end_raw = MIN(offset + length, o->onode.size);
-    uint64_t cow_end = ROUND_UP_TO(cow_end_raw, block_size);
-    bp = o->onode.find_lextent(cow_offset_raw);
-    bluestore_blob_t *b = o->get_blob_ptr(bp->second.blob);
-    if (cow_end > cow_offset) {
-      uint64_t cow_length = cow_end - cow_offset;
-      uint64_t x_off = cow_offset - bp->first;
-      dout(20) << "  tail shared, will COW 0x"
-	       << std::hex << x_off << "~0x" << cow_length << std::dec
-	       << " from " << bp->second
-	       << " blob " << *b << dendl;
-      while (cow_length > 0) {
-	cow_tail_op = _get_wal_op(txc, o);
-	cow_tail_op->op = bluestore_wal_op_t::OP_COPY;
-	uint64_t c_len;
-	cow_tail_op->src_extent.offset =
-	  b->calc_offset(x_off + bp->second.offset, &c_len);
-	c_len = MIN(c_len, cow_length);
-	cow_tail_op->src_extent.length = c_len;
-	// will adjust logical offset -> final bdev offset below
-	cow_tail_op->extent.offset = cow_offset;
-	cow_tail_op->extent.length = c_len;
-	cow_offset += c_len;
-	cow_length -= c_len;
-	x_off += c_len;
-	assert(cow_length == 0);  // fixme ... until we fix this loop
+      if (tail_read) {
+	bufferlist tail_bl;
+	int r = _do_read(c.get(), o, offset + length + tail_pad, tail_read,
+			 tail_bl, 0);
+	assert(r >= 0 && r <= (int)tail_read);
+	b_len += r;
```

Comment:
Shouldn't that be 
b_len += tail_read; 
to have b_len chunk_aligned and synced with padded.length()?